File: CSharp/JacobianTests.cs
using System;
using NUnit.Framework;

[TestFixture]
public class JacobianTests
{
    [Test]
    public void TestGetJacobianF()
    {
        double[] x = { 1.0, 2.0 };
        double[,] jacobian = JacobianCalculator.GetJacobian(Functions.F, x, 2);
        double[,] expected = { { 2.0, 0.0 }, { 0.0, 12.0 } };

        Assert.IsTrue(IsApproximatelyEqual(jacobian, expected, 1e-5));
    }

    [Test]
    public void TestGetJacobianG()
    {
        double twoPi = 2 * Math.PI;
        for (double theta_a = 0; theta_a <= twoPi; theta_a += Math.PI / 50)
        {
            for (double theta_b = 0; theta_b <= twoPi; theta_b += Math.PI / 50)
            {
                double[] x = { theta_a, theta_b };
                double[,] jacobian = JacobianCalculator.GetJacobian(Functions.G, x, 2);
                double[,] expected = {
                    { Math.Cos(x[1]) * Math.Cos(x[0]), -Math.Sin(x[1]) * Math.Sin(x[0]) },
                    { Math.Cos(x[0]), 0.0 }
                };

                Assert.IsTrue(IsApproximatelyEqual(jacobian, expected, 1e-5), 
                    $"Test case G failed for theta_a={theta_a}, theta_b={theta_b}");
            }
        }
    }

    private bool IsApproximatelyEqual(double[,] a, double[,] b, double tolerance)
    {
        int rows = a.GetLength(0);
        int cols = a.GetLength(1);

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                if (Math.Abs(a[i, j] - b[i, j]) > tolerance)
                {
                    return false;
                }
            }
        }

        return true;
    }
}

File: java/JacobianTests.java
import static org.junit.jupiter.api.Assertions.assertArrayEquals;

import org.junit.jupiter.api.Test;

public class JacobianTests {

    @Test
    public void testGetJacobianF() {
        double[] x = {1.0, 2.0};
        double[][] jacobian = JacobianCalculator.getJacobian(Functions::f, x, 2);
        double[][] expected = {
            {2.0, 0.0},
            {0.0, 12.0}
        };

        assertArrayEquals(expected, jacobian);
    }

    @Test
    public void testGetJacobianG() {
        double twoPi = 2 * Math.PI;
        for (double theta_a = 0; theta_a <= twoPi; theta_a += Math.PI / 50) {
            for (double theta_b = 0; theta_b <= twoPi; theta_b += Math.PI / 50) {
                double[] x = {theta_a, theta_b};
                double[][] jacobian = JacobianCalculator.getJacobian(Functions::g, x, 2);
                double[][] expected = {
                    {Math.cos(x[1]) * Math.cos(x[0]), -Math.sin(x[1]) * Math.sin(x[0])},
                    {Math.cos(x[0]), 0.0}
                };

                assertArrayEquals(expected, jacobian, 1e-5);
            }
        }
    }
}
